<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **SpaceX  Falcon 9 first stage Landing Prediction**


# Lab 1: Collecting the data


Estimated time needed: **45** minutes


In this capstone, we will predict if the Falcon 9 first stage will land successfully. SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars; other providers cost upward of 165 million dollars each, much of the savings is because SpaceX can reuse the first stage. Therefore if we can determine if the first stage will land, we can determine the cost of a launch. This information can be used if an alternate company wants to bid against SpaceX for a rocket launch. In this lab, you will collect and make sure the data is in the correct format from an API. The following is an example of a successful and launch.


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/landing\_1.gif)


Several examples of an unsuccessful landing are shown here:


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/crash.gif)


Most unsuccessful landings are planned. Space X performs a controlled landing in the oceans.


## Objectives


In this lab, you will make a get request to the SpaceX API. You will also do some basic data wrangling and formating.

*   Request to the SpaceX API
*   Clean the requested data


***


## Import Libraries and Define Auxiliary Functions


We will import the following libraries into the lab


In [1]:
# Requests allows us to make HTTP requests which we will use to get data from an API
import requests
# Pandas is a software library written for the Python programming language for data manipulation and analysis.
import pandas as pd
# NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays
import numpy as np
# Datetime is a library that allows us to represent dates
import datetime

# Setting this option will print all collumns of a dataframe
pd.set_option('display.max_columns', None)
# Setting this option will print all of the data in a feature
pd.set_option('display.max_colwidth', None)

Below we will define a series of helper functions that will help us use the API to extract information using identification numbers in the launch data.

From the <code>rocket</code> column we would like to learn the booster name.


In [2]:
# Takes the dataset and uses the rocket column to call the API and append the data to the list
def getBoosterVersion(data):
    for x in data['rocket']:
        response = requests.get("https://api.spacexdata.com/v4/rockets/"+str(x)).json()
        BoosterVersion.append(response['name'])

From the <code>launchpad</code> we would like to know the name of the launch site being used, the logitude, and the latitude.


In [3]:
# Takes the dataset and uses the launchpad column to call the API and append the data to the list
def getLaunchSite(data):
    for x in data['launchpad']:
        response = requests.get("https://api.spacexdata.com/v4/launchpads/"+str(x)).json()
        Longitude.append(response['longitude'])
        Latitude.append(response['latitude'])
        LaunchSite.append(response['name'])

From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to.


In [4]:
# Takes the dataset and uses the payloads column to call the API and append the data to the lists
def getPayloadData(data):
    for load in data['payloads']:
        response = requests.get("https://api.spacexdata.com/v4/payloads/"+load).json()
        PayloadMass.append(response['mass_kg'])
        Orbit.append(response['orbit'])

From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, wheter the core is reused, wheter legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.


In [5]:
# Takes the dataset and uses the cores column to call the API and append the data to the lists
def getCoreData(data):
    for core in data['cores']:
            if core['core'] != None:
                response = requests.get("https://api.spacexdata.com/v4/cores/"+core['core']).json()
                Block.append(response['block'])
                ReusedCount.append(response['reuse_count'])
                Serial.append(response['serial'])
            else:
                Block.append(None)
                ReusedCount.append(None)
                Serial.append(None)
            Outcome.append(str(core['landing_success'])+' '+str(core['landing_type']))
            Flights.append(core['flight'])
            GridFins.append(core['gridfins'])
            Reused.append(core['reused'])
            Legs.append(core['legs'])
            LandingPad.append(core['landpad'])

Now let's start requesting rocket launch data from SpaceX API with the following URL:


In [6]:
spacex_url="https://api.spacexdata.com/v4/launches/past"

In [7]:
response = requests.get(spacex_url)

Check the content of the response


In [ ]:
print(response.content)

You should see the response contains massive information about SpaceX launches. Next, let's try to discover some more relevant information for this project.


### Task 1: Request and parse the SpaceX launch data using the GET request


To make the requested JSON results more consistent, we will use the following static response object for this project:


In [8]:
static_json_url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/API_call_spacex_api.json'

We should see that the request was successfull with the 200 status response code


In [9]:
response.status_code

200

Now we decode the response content as a Json using <code>.json()</code> and turn it into a Pandas dataframe using <code>.json_normalize()</code>


In [10]:
# Use json_normalize meethod to convert the json result into a dataframe
import json
data_json = response.json()
data = pd.json_normalize(data_json)

Using the dataframe <code>data</code> print the first 5 rows


In [11]:
# Get the head of the dataframe
data.head()

static_fire_date_utc  static_fire_date_unix    net  window  \
0  2006-03-17T00:00:00.000Z           1.142554e+09  False     0.0   
1                      None                    NaN  False     0.0   
2                      None                    NaN  False     0.0   
3  2008-09-20T00:00:00.000Z           1.221869e+09  False     0.0   
4                      None                    NaN  False     0.0   

                     rocket  success  \
0  5e9d0d95eda69955f709d1eb    False   
1  5e9d0d95eda69955f709d1eb    False   
2  5e9d0d95eda69955f709d1eb    False   
3  5e9d0d95eda69955f709d1eb     True   
4  5e9d0d95eda69955f709d1eb     True   

                                                                                                            failures  \
0                                                [{'time': 33, 'altitude': None, 'reason': 'merlin engine failure'}]   
1            [{'time': 301, 'altitude': 289, 'reason': 'harmonic oscillation leading to premature engine shutdown'}]   
2  [{'time': 140, 'altitude': 35, 'reason': 'residual stage-1 thrust led to collision between stage 1 and stage 2'}]   
3                                                                                                                 []   
4                                                                                                                 []   

                                                                                                                                                                                details  \
0                                                                                                                                      Engine failure at 33 seconds and loss of vehicle   
1  Successful first stage burn and transition to second stage, maximum altitude 289 km, Premature engine shutdown at T+7 min 30 s, Failed to reach orbit, Failed to recover first stage   
2                                                                                                                  Residual stage 1 thrust led to collision between stage 1 and stage 2   
3                        Ratsat was carried to orbit on the first successful orbital launch of any privately funded and developed, liquid-propelled carrier rocket, the SpaceX Falcon 1   
4                                                                                                                                                                                  None   

  crew ships capsules                                              payloads  \
0   []    []       []                            [5eb0e4b5b6c3bb0006eeb1e1]   
1   []    []       []                            [5eb0e4b6b6c3bb0006eeb1e2]   
2   []    []       []  [5eb0e4b6b6c3bb0006eeb1e3, 5eb0e4b6b6c3bb0006eeb1e4]   
3   []    []       []                            [5eb0e4b7b6c3bb0006eeb1e5]   
4   []    []       []                            [5eb0e4b7b6c3bb0006eeb1e6]   

                  launchpad  flight_number         name  \
0  5e9e4502f5090995de566f86              1    FalconSat   
1  5e9e4502f5090995de566f86              2      DemoSat   
2  5e9e4502f5090995de566f86              3  Trailblazer   
3  5e9e4502f5090995de566f86              4       RatSat   
4  5e9e4502f5090995de566f86              5     RazakSat   

                   date_utc   date_unix                 date_local  \
0  2006-03-24T22:30:00.000Z  1143239400  2006-03-25T10:30:00+12:00   
1  2007-03-21T01:10:00.000Z  1174439400  2007-03-21T13:10:00+12:00   
2  2008-08-03T03:34:00.000Z  1217734440  2008-08-03T15:34:00+12:00   
3  2008-09-28T23:15:00.000Z  1222643700  2008-09-28T11:15:00+12:00   
4  2009-07-13T03:35:00.000Z  1247456100  2009-07-13T15:35:00+12:00   

  date_precision  upcoming  \
0           hour     False   
1           hour     False   
2           hour     False   
3           hour     False   
4           hour     False   

                                                                           

You will notice that a lot of the data are IDs. For example the rocket column has no information about the rocket just an identification number.

We will now use the API again to get information about the launches using the IDs given for each launch. Specifically we will be using columns <code>rocket</code>, <code>payloads</code>, <code>launchpad</code>, and <code>cores</code>.


In [12]:
# Lets take a subset of our dataframe keeping only the features we want and the flight number, and date_utc.
data = data[['rocket', 'payloads', 'launchpad', 'cores', 'flight_number', 'date_utc']]

In [13]:
data.head()
data_t= data[data['payloads'].map(len)>2]
data_t.head()

rocket  \
74   5e9d0d95eda69973a809d1ec   
80   5e9d0d95eda69974db09d1ed   
126  5e9d0d95eda69973a809d1ec   

                                                                                                                                                                                                                                                                                                                                                                                                                             payloads  \
74                                                                                                                                                                                                                                                                                                                                                     [5eb0e4cab6c3bb0006eeb232, 5eb0e4cab6c3bb0006eeb233, 5eb0e4cab6c3bb0006eeb234]   
80   [5eb0e4ccb6c3bb0006eeb23a, 5eb0e4ccb6c3bb0006eeb23b, 5eb0e4ccb6c3bb0006eeb23c, 5eb0e4ccb6c3bb0006eeb23d, 5eb0e4ccb6c3bb0006eeb23e, 5eb0e4cdb6c3bb0006eeb23f, 5eb0e4cdb6c3bb0006eeb240, 5eb0e4cdb6c3bb0006eeb241, 5eb0e4cdb6c3bb0006eeb242, 5eb0e4cdb6c3bb0006eeb243, 5eb0e4cdb6c3bb0006eeb244, 5eb0e4cdb6c3bb0006eeb245, 5eb0e4ceb6c3bb0006eeb246, 5eb0e4ceb6c3bb0006eeb247, 5eb0e4ceb6c3bb0006eeb248, 5eb0e4ceb6c3bb0006eeb249]   
126                                                                                                                                                                                                                                                                                                                                                    [605b4bfcaa5433645e37d048, 609f48374a12e4692eae4667, 609f49c64a12e4692eae4668]   

                    launchpad  \
74   5e9e4501f509094ba4566f84   
80   5e9e4502f509094188566f88   
126  5e9e4502f509094188566f88   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              cores  \
74                                                                                                                                                                                                                                                                                                                                                                                                                                             [{'core': '5e9e28a5f3591809c03b2658', 'flight': 3, 'gridfins': True, 'legs': True, 'reused': True, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'ASDS', 'landpad': '5e9e3032383ecb6bb234e7ca'}]   
80   [{'core': '5e9e28a7f3591878063b2661', 'flight': 1, 'gridfins': True, 'legs': True, 'reused': False, 'landing_attempt': True, 'landing_success': False, 'landing_type': 'ASDS', 'landpad': '5e9e3032383ecb6bb234e7ca'}, {'core': '5e9e28a6f359183c413b265d', 'flight': 2, 'gridfins': True, 'legs': True, 'reused': True, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'RTLS', 'landpad': '5e9e3032383ecb267a34e7c7'}, {'core': '5e9e28a6f359188fd53b265e', 'flight': 2, 'gridfins': True, 'legs': True, 'reused': True, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'RTLS', 'landpad': '5e9e3032383ecb90a834e7c8'}]   
126                                                                                                                

In [14]:
# We will remove rows with multiple cores because those are falcon rockets with 2 extra rocket boosters and rows that have multiple payloads in a single rocket.
data = data[data['cores'].map(len)==1]
data = data[data['payloads'].map(len)==1]

In [15]:
data.head()

rocket                    payloads  \
0  5e9d0d95eda69955f709d1eb  [5eb0e4b5b6c3bb0006eeb1e1]   
1  5e9d0d95eda69955f709d1eb  [5eb0e4b6b6c3bb0006eeb1e2]   
3  5e9d0d95eda69955f709d1eb  [5eb0e4b7b6c3bb0006eeb1e5]   
4  5e9d0d95eda69955f709d1eb  [5eb0e4b7b6c3bb0006eeb1e6]   
5  5e9d0d95eda69973a809d1ec  [5eb0e4b7b6c3bb0006eeb1e7]   

                  launchpad  \
0  5e9e4502f5090995de566f86   
1  5e9e4502f5090995de566f86   
3  5e9e4502f5090995de566f86   
4  5e9e4502f5090995de566f86   
5  5e9e4501f509094ba4566f84   

                                                                                                                                                                                              cores  \
0  [{'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   
1  [{'core': '5e9e289ef35918416a3b2624', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   
3  [{'core': '5e9e289ef3591855dc3b2626', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   
4  [{'core': '5e9e289ef359184f103b2627', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   
5  [{'core': '5e9e289ef359185f2b3b2628', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   

   flight_number                  date_utc  
0              1  2006-03-24T22:30:00.000Z  
1              2  2007-03-21T01:10:00.000Z  
3              4  2008-09-28T23:15:00.000Z  
4              5  2009-07-13T03:35:00.000Z  
5              6  2010-06-04T18:45:00.000Z

In [16]:
# Since payloads and cores are lists of size 1 we will also extract the single value in the list and replace the feature.
data['cores'] = data['cores'].map(lambda x : x[0])
data['payloads'] = data['payloads'].map(lambda x : x[0])

In [17]:
data.head()

rocket                  payloads  \
0  5e9d0d95eda69955f709d1eb  5eb0e4b5b6c3bb0006eeb1e1   
1  5e9d0d95eda69955f709d1eb  5eb0e4b6b6c3bb0006eeb1e2   
3  5e9d0d95eda69955f709d1eb  5eb0e4b7b6c3bb0006eeb1e5   
4  5e9d0d95eda69955f709d1eb  5eb0e4b7b6c3bb0006eeb1e6   
5  5e9d0d95eda69973a809d1ec  5eb0e4b7b6c3bb0006eeb1e7   

                  launchpad  \
0  5e9e4502f5090995de566f86   
1  5e9e4502f5090995de566f86   
3  5e9e4502f5090995de566f86   
4  5e9e4502f5090995de566f86   
5  5e9e4501f509094ba4566f84   

                                                                                                                                                                                            cores  \
0  {'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
1  {'core': '5e9e289ef35918416a3b2624', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
3  {'core': '5e9e289ef3591855dc3b2626', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
4  {'core': '5e9e289ef359184f103b2627', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
5  {'core': '5e9e289ef359185f2b3b2628', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   

   flight_number                  date_utc  
0              1  2006-03-24T22:30:00.000Z  
1              2  2007-03-21T01:10:00.000Z  
3              4  2008-09-28T23:15:00.000Z  
4              5  2009-07-13T03:35:00.000Z  
5              6  2010-06-04T18:45:00.000Z

In [18]:
data.dtypes

rocket           object
payloads         object
launchpad        object
cores            object
flight_number     int64
date_utc         object
dtype: object

In [19]:
# We also want to convert the date_utc to a datetime datatype and then extracting the date leaving the time
data['date'] = pd.to_datetime(data['date_utc']).dt.date

# Using the date we will restrict the dates of the launches
data = data[data['date'] <= datetime.date(2020, 11, 13)]

*   From the <code>rocket</code> we would like to learn the booster name

*   From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to

*   From the <code>launchpad</code> we would like to know the name of the launch site being used, the longitude, and the latitude.

*   From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, whether the core is reused, whether legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.

The data from these requests will be stored in lists and will be used to create a new dataframe.


In [20]:
#Global variables 
BoosterVersion = []
PayloadMass = []
Orbit = []
LaunchSite = []
Outcome = []
Flights = []
GridFins = []
Reused = []
Legs = []
LandingPad = []
Block = []
ReusedCount = []
Serial = []
Longitude = []
Latitude = []

These functions will apply the outputs globally to the above variables. Let's take a looks at <code>BoosterVersion</code> variable. Before we apply  <code>getBoosterVersion</code> the list is empty:


In [21]:
BoosterVersion

[]

Now, let's apply <code> getBoosterVersion</code> function method to get the booster version


In [22]:
# Call getBoosterVersion
getBoosterVersion(data)

the list has now been update


In [23]:
BoosterVersion[0:5]

['Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 9']

In [24]:
len(BoosterVersion)

94

we can apply the rest of the  functions here:


In [25]:
# Call getLaunchSite
getLaunchSite(data)
LaunchSite[0:5]

['Kwajalein Atoll',
 'Kwajalein Atoll',
 'Kwajalein Atoll',
 'Kwajalein Atoll',
 'CCSFS SLC 40']

In [26]:
# Call getPayloadData
getPayloadData(data)
PayloadMass[0:5]

[20, None, 165, 200, None]

In [27]:
# Call getCoreData
getCoreData(data)

In [28]:
data.head()

rocket                  payloads  \
0  5e9d0d95eda69955f709d1eb  5eb0e4b5b6c3bb0006eeb1e1   
1  5e9d0d95eda69955f709d1eb  5eb0e4b6b6c3bb0006eeb1e2   
3  5e9d0d95eda69955f709d1eb  5eb0e4b7b6c3bb0006eeb1e5   
4  5e9d0d95eda69955f709d1eb  5eb0e4b7b6c3bb0006eeb1e6   
5  5e9d0d95eda69973a809d1ec  5eb0e4b7b6c3bb0006eeb1e7   

                  launchpad  \
0  5e9e4502f5090995de566f86   
1  5e9e4502f5090995de566f86   
3  5e9e4502f5090995de566f86   
4  5e9e4502f5090995de566f86   
5  5e9e4501f509094ba4566f84   

                                                                                                                                                                                            cores  \
0  {'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
1  {'core': '5e9e289ef35918416a3b2624', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
3  {'core': '5e9e289ef3591855dc3b2626', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
4  {'core': '5e9e289ef359184f103b2627', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   
5  {'core': '5e9e289ef359185f2b3b2628', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   

   flight_number                  date_utc        date  
0              1  2006-03-24T22:30:00.000Z  2006-03-24  
1              2  2007-03-21T01:10:00.000Z  2007-03-21  
3              4  2008-09-28T23:15:00.000Z  2008-09-28  
4              5  2009-07-13T03:35:00.000Z  2009-07-13  
5              6  2010-06-04T18:45:00.000Z  2010-06-04

Finally lets construct our dataset using the data we have obtained. We we combine the columns into a dictionary.


In [29]:
launch_dict = {'FlightNumber': list(data['flight_number']),
'Date': list(data['date']),
'BoosterVersion':BoosterVersion,
'PayloadMass':PayloadMass,
'Orbit':Orbit,
'LaunchSite':LaunchSite,
'Outcome':Outcome,
'Flights':Flights,
'GridFins':GridFins,
'Reused':Reused,
'Legs':Legs,
'LandingPad':LandingPad,
'Block':Block,
'ReusedCount':ReusedCount,
'Serial':Serial,
'Longitude': Longitude,
'Latitude': Latitude}


Then, we need to create a Pandas data frame from the dictionary launch_dict.


In [30]:
# Create a data from launch_dict
df=pd.DataFrame.from_dict(launch_dict, orient='index')

Show the summary of the dataframe


In [31]:
# Show the head of the dataframe
df.head()

0           1           2           3           4   \
FlightNumber             1           2           4           5           6   
Date            2006-03-24  2007-03-21  2008-09-28  2009-07-13  2010-06-04   
BoosterVersion    Falcon 1    Falcon 1    Falcon 1    Falcon 1    Falcon 9   
PayloadMass             20        None         165         200        None   
Orbit                  LEO         LEO         LEO         LEO         LEO   

                        5           6           7           8           9   \
FlightNumber             8          10          11          12          13   
Date            2012-05-22  2013-03-01  2013-09-29  2013-12-03  2014-01-06   
BoosterVersion    Falcon 9    Falcon 9    Falcon 9    Falcon 9    Falcon 9   
PayloadMass            525         677         500        3170        3325   
Orbit                  LEO         ISS          PO         GTO         GTO   

                        10          11          12          13          14  \
FlightNumber            14          15          16          17          18   
Date            2014-04-18  2014-07-14  2014-08-05  2014-09-07  2014-09-21   
BoosterVersion    Falcon 9    Falcon 9    Falcon 9    Falcon 9    Falcon 9   
PayloadMass           2296        1316        4535        4428        2216   
Orbit                  ISS         LEO         GTO         GTO         ISS   

                        15          16          17          18          19  \
FlightNumber            19          20          22          23          24   
Date            2015-01-10  2015-02-11  2015-04-14  2015-04-27  2015-06-28   
BoosterVersion    Falcon 9    Falcon 9    Falcon 9    Falcon 9    Falcon 9   
PayloadMass           2395         570        1898        4707        2477   
Orbit                  ISS       ES-L1         ISS         GTO         ISS   

                        20          21          22          23          24  \
FlightNumber            25          26          27          28          29   
Date            2015-12-22  2016-01-17  2016-03-04  2016-04-08  2016-05-06   
BoosterVersion    Falcon 9    Falcon 9    Falcon 9    Falcon 9    Falcon 9   
PayloadMass           2034         553        5271        3136        4696   
Orbit                  LEO          PO         GTO         ISS         GTO   

                        25          26          27          28          29  \
FlightNumber            30          32          33          34          35   
Date            2016-05-27  2016-07-18  2016-08-14  2016-09-01  2017-01-14   
BoosterVersion    Falcon 9    Falcon 9    Falcon 9    Falcon 9    Falcon 9   
PayloadMass           3100        2257        4600        5500        9600   
Orbit                  GTO         ISS         GTO         GTO          PO   

                        30          31          32          33          34  \
FlightNumber            36          37          38          39          40   
Date            2017-02-19  2017-03-16  2017-03-30  2017-05-01  2017-05-15   
BoosterVersion    Falcon 9    Falcon 9    Falcon 9    Falcon 9    Falcon 9   
PayloadMass           2490        5600        5300        None        6070   
Orbit                  ISS         GTO         GTO         LEO         GTO   

                        35          36          37          38          39  \
FlightNumber            41          42          43          44          45   
Date            2017-06-03  2017-06-23  2017-06-25  2017-07-05  2017-08-14   
BoosterVersion    Falcon 9    Falcon 9    Falcon 9    Falcon 9    Falcon 9   
PayloadMass           2708        3669        9600        6761        2910   
Orbit                  ISS         GTO          PO         GTO         ISS   

                        40          41          42          43          44  \
FlightNumber            46          47          48          49          50   
Date            2017-08-24  2017-09-07  2017-10-09  2017-10-11  2017-10-30   
BoosterVersion    Falcon 9    Falcon 9

In [32]:
df = df.transpose()

In [34]:
df.isnull().sum()

FlightNumber       0
Date               0
BoosterVersion     0
PayloadMass        6
Orbit              0
LaunchSite         0
Outcome            0
Flights            0
GridFins           0
Reused             0
Legs               0
LandingPad        30
Block              4
ReusedCount        0
Serial             0
Longitude          0
Latitude           0
dtype: int64

### Task 2: Filter the dataframe to only include `Falcon 9` launches


Finally we will remove the Falcon 1 launches keeping only the Falcon 9 launches. Filter the data dataframe using the <code>BoosterVersion</code> column to only keep the Falcon 9 launches. Save the filtered data to a new dataframe called <code>data_falcon9</code>.


In [35]:
# Hint data['BoosterVersion']!='Falcon 1'
df[['BoosterVersion']].value_counts()

BoosterVersion
Falcon 9          90
Falcon 1           4
dtype: int64

In [36]:
data_falcon9=df[df['BoosterVersion']!='Falcon 1']

In [37]:
data_falcon9.shape

(90, 17)

In [38]:
data_falcon9.head()

FlightNumber        Date BoosterVersion PayloadMass Orbit    LaunchSite  \
4            6  2010-06-04       Falcon 9        None   LEO  CCSFS SLC 40   
5            8  2012-05-22       Falcon 9         525   LEO  CCSFS SLC 40   
6           10  2013-03-01       Falcon 9         677   ISS  CCSFS SLC 40   
7           11  2013-09-29       Falcon 9         500    PO   VAFB SLC 4E   
8           12  2013-12-03       Falcon 9        3170   GTO  CCSFS SLC 40   

       Outcome Flights GridFins Reused   Legs LandingPad Block ReusedCount  \
4    None None       1    False  False  False       None     1           0   
5    None None       1    False  False  False       None     1           0   
6    None None       1    False  False  False       None     1           0   
7  False Ocean       1    False  False  False       None     1           0   
8    None None       1    False  False  False       None     1           0   

  Serial   Longitude   Latitude  
4  B0003  -80.577366  28.561857  
5  B0005  -80.577366  28.561857  
6  B0007  -80.577366  28.561857  
7  B1003 -120.610829  34.632093  
8  B1004  -80.577366  28.561857

Now that we have removed some values we should reset the FlgihtNumber column


In [39]:
data_falcon9.loc[:,'FlightNumber'] = list(range(1, data_falcon9.shape[0]+1))
data_falcon9

C:\Users\80302017\Anaconda3\lib\site-packages\pandas\core\indexing.py:1951: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[selected_item_labels] = value


FlightNumber        Date BoosterVersion PayloadMass Orbit    LaunchSite  \
4              1  2010-06-04       Falcon 9        None   LEO  CCSFS SLC 40   
5              2  2012-05-22       Falcon 9         525   LEO  CCSFS SLC 40   
6              3  2013-03-01       Falcon 9         677   ISS  CCSFS SLC 40   
7              4  2013-09-29       Falcon 9         500    PO   VAFB SLC 4E   
8              5  2013-12-03       Falcon 9        3170   GTO  CCSFS SLC 40   
..           ...         ...            ...         ...   ...           ...   
89            86  2020-09-03       Falcon 9       15600  VLEO    KSC LC 39A   
90            87  2020-10-06       Falcon 9       15600  VLEO    KSC LC 39A   
91            88  2020-10-18       Falcon 9       15600  VLEO    KSC LC 39A   
92            89  2020-10-24       Falcon 9       15600  VLEO  CCSFS SLC 40   
93            90  2020-11-05       Falcon 9        3681   MEO  CCSFS SLC 40   

        Outcome Flights GridFins Reused   Legs                LandingPad  \
4     None None       1    False  False  False                      None   
5     None None       1    False  False  False                      None   
6     None None       1    False  False  False                      None   
7   False Ocean       1    False  False  False                      None   
8     None None       1    False  False  False                      None   
..          ...     ...      ...    ...    ...                       ...   
89    True ASDS       2     True   True   True  5e9e3032383ecb6bb234e7ca   
90    True ASDS       3     True   True   True  5e9e3032383ecb6bb234e7ca   
91    True ASDS       6     True   True   True  5e9e3032383ecb6bb234e7ca   
92    True ASDS       3     True   True   True  5e9e3033383ecbb9e534e7cc   
93    True ASDS       1     True  False   True  5e9e3032383ecb6bb234e7ca   

   Block ReusedCount Serial   Longitude   Latitude  
4      1           0  B0003  -80.577366  28.561857  
5      1           0  B0005  -80.577366  28.561857  
6      1           0  B0007  -80.577366  28.561857  
7      1           0  B1003 -120.610829  34.632093  
8      1           0  B1004  -80.577366  28.561857  
..   ...         ...    ...         ...        ...  
89     5          12  B1060  -80.603956  28.608058  
90     5          12  B1058  -80.603956  28.608058  
91     5          11  B1051  -80.603956  28.608058  
92     5          12  B1060  -80.577366  28.561857  
93     5           6  B1062  -80.577366  28.561857  

[90 rows x 17 columns]

## Data Wrangling


We can see below that some of the rows are missing values in our dataset.


In [40]:
data_falcon9.isnull().sum()

FlightNumber       0
Date               0
BoosterVersion     0
PayloadMass        5
Orbit              0
LaunchSite         0
Outcome            0
Flights            0
GridFins           0
Reused             0
Legs               0
LandingPad        26
Block              0
ReusedCount        0
Serial             0
Longitude          0
Latitude           0
dtype: int64

Before we can continue we must deal with these missing values. The <code>LandingPad</code> column will retain None values to represent when landing pads were not used.


### Task 3: Dealing with Missing Values


Calculate below the mean for the <code>PayloadMass</code> using the <code>.mean()</code>. Then use the mean and the <code>.replace()</code> function to replace `np.nan` values in the data with the mean you calculated.


In [41]:
# Calculate the mean value of PayloadMass column
payload_mass_mean = data_falcon9['PayloadMass'].mean()
payload_mass_mean
# Replace the np.nan values with its mean value
data_falcon9['PayloadMass'].fillna(value=payload_mass_mean, inplace=True)

C:\Users\80302017\Anaconda3\lib\site-packages\pandas\core\generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [42]:
data_falcon9.isnull().sum()

FlightNumber       0
Date               0
BoosterVersion     0
PayloadMass        0
Orbit              0
LaunchSite         0
Outcome            0
Flights            0
GridFins           0
Reused             0
Legs               0
LandingPad        26
Block              0
ReusedCount        0
Serial             0
Longitude          0
Latitude           0
dtype: int64

You should see the number of missing values of the <code>PayLoadMass</code> change to zero.


Now we should have no missing values in our dataset except for in <code>LandingPad</code>.


We can now export it to a <b>CSV</b> for the next section,but to make the answers consistent, in the next lab we will provide data in a pre-selected date range.


<code>data_falcon9.to_csv('dataset_part\_1.csv', index=False)</code>

In [43]:
data_falcon9.to_csv('dataset_part_1.csv', index=False)

## Authors


<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2022-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By | Change Description                  |
| ----------------- | ------- | ---------- | ----------------------------------- |
| 2020-09-20        | 1.1     | Joseph     | get result each time you run        |
| 2020-09-20        | 1.1     | Azim       | Created Part 1 Lab using SpaceX API |
| 2020-09-20        | 1.0     | Joseph     | Modified Multiple Areas             |


Copyright © 2021 IBM Corporation. All rights reserved.
